lib/python/qmk/cli/c2json.py

```diff
@@ -0,0 +1,324 @@
+    ...          ))
+    (['LALT', 'LSFT', 'LCTL'], 'KC_A')
+    """
+
+    log.debug('handling operation AST: %r', or_ast)
+    queue = [or_ast]
+    keycodes = []
+    # stage 1: traverse the tree and pull out all the keycodes
+    while queue:
+        item = queue.pop()
+        if type(item) == ID:
+            keycodes.append(item.name)
+        elif type(item) == BinaryOp:
+            if item.op != '|':
+                log.warn('Encountered non-OR binary operation in your keymap! %r', item)
+                continue
+            queue.append(item.left)
+            queue.append(item.right)
+
+    # stage 2: separate these into modifiers and keycodes
+    modifiers = []
+    keycode = None
+    for kc in keycodes:
+        if kc in MODIFIER_MAP:
+            modifiers.append(MODIFIER_MAP[kc])
+        else:
+            keycode = kc
+    return (modifiers, keycode)
+
+
+def make_modifier_keycode(modifiers, keycode):
+    """
+    Turns a list of modifiers into parenthesized form
+
+    Parameters:
+    modifiers -- List[str] of modifiers
+    keycode -- string keycode to apply them to
+
+    Example:
+    >>> make_modifier_keycode(['LALT', 'LSFT', 'LCTL'], 'KC_A')
+    'LCTL(LSFT(LALT(KC_A)))'
+    """
+    result = keycode
+    for modifier in modifiers:
+        result = '{}({})'.format(modifier, result)
+    return result
+
+
+def extract_row(row_ast, log):
+    """
+    Extracts a row as a List[Key] of keycode strings
+
+    Transforms an InitList of keycode ASTs into a list of string QMK keycodes
+    """
+    # each layer is a
+    # InitList of rows ->
+    #   InitList of IDs or more complex objects
+    row = []
+    for expr in row_ast.exprs:
+        if type(expr) == ID:
+            row.append(expr.name)
+        elif type(expr) == Constant and expr.type == 'int':
+            if expr.value != '0x01':
+                # KC_TRNS is represented by 0x01
+                log.warn('Discarding an unrecognized numeric keycode: %s', expr.value)
+            row.append('KC_TRNS')
+        elif type(expr) == BinaryOp:
+            # it's probably a modifier set
+            modifiers, keycode = or_ast_to_modifiers(expr, log=log)
+            row.append(make_modifier_keycode(modifiers, keycode))
+        else:
+            log.warn('Not implemented: emitting KC_TRNS for %s', expr)
+            row.append('KC_TRNS')
+            continue
+    return row
+
+
+def extract_layers(keymaps, log):
+    """
+    Takes an AST representation of the keymaps array and pulls the layers out
+    as Layer objects
+
+    Returns:
+    List[Layer] representing the layers in the AST
+    """
+    layers = []
+    for layer in keymaps.init.exprs:
+        name = layer.name[0].value
+        rows_ast = layer.expr.exprs
+        rows = []
+        for row_ast in rows_ast:
+            rows.append(extract_row(row_ast, log=log))
+        layers.append(Layer(name=name, keys=rows))
+    return layers
+
+
+def find_keyboard_name_from_keymap_c(keymap_c):
```

Comment:
Is a keyboard name always required? What happens if someone is trying to convert a keymap in the `qmk_firmware/layouts` directory? We will also soon support keymaps in userspace.